File: im-desktop/services/botService.ts
/**
 * AI 聊天机器人服务 - TypeScript API 封装
 * 支持 OpenAI/Claude/Gemini/Custom Webhook 多平台 AI 模型
 */

const API_BASE = '/api/bots';

export interface Bot {
  botId: string;
  name: string;
  description: string;
  avatarUrl?: string;
  ownerId: string;
  botType: 'OPENAI' | 'CLAUDE' | 'GEMINI' | 'CUSTOM' | 'LOCAL';
  modelName: string;
  apiKey?: string;
  apiBaseUrl?: string;
  webhookUrl?: string;
  webhookSecret?: string;
  systemPrompt?: string;
  maxTokens: number;
  temperature: number;
  status: 'ACTIVE' | 'INACTIVE' | 'DELETED';
  isPublic: boolean;
  enableImageGen: boolean;
  enableSpeechToText: boolean;
  rateLimit: number;
  sessionCount: number;
  messageCount: number;
  totalTokensUsed: number;
  accessToken?: string;
  createdAt: string;
  updatedAt: string;
  lastActiveAt?: string;
}

export interface BotSession {
  sessionId: string;
  botId: string;
  userId: string;
  conversationId: string;
  contextTokens: number;
  turnCount: number;
  totalTokensUsed: number;
  status: 'ACTIVE' | 'ENDED' | 'TIMEOUT';
  endReason?: string;
  promptVersion?: string;
  createdAt: string;
  lastMessageAt: string;
  endedAt?: string;
}

export interface ChatResponse {
  reply: string;
  botId: string;
  timestamp: string;
}

export interface BotStats {
  totalSessions: number;
  activeSessions: number;
  totalTokens: number;
  avgTurns: number;
  totalMessages: number;
  totalTokensUsed: number;
}

export interface BotPreset {
  type: string;
  model: string;
  name: string;
}

// ========== Bot CRUD ==========

export async function createBot(data: {
  name: string;
  description?: string;
  botType?: string;
  modelName?: string;
  ownerId: string;
}): Promise<Bot> {
  const res = await fetch(`${API_BASE}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  });
  if (!res.ok) throw new Error(`创建机器人失败: ${res.status}`);
  return res.json();
}

export async function getBot(botId: string): Promise<Bot> {
  const res = await fetch(`${API_BASE}/${botId}`);
  if (!res.ok) throw new Error(`获取机器人失败: ${res.status}`);
  return res.json();
}

export async function getBotByToken(token: string): Promise<Bot> {
  const res = await fetch(`${API_BASE}/token/${token}`);
  if (!res.ok) throw new Error(`Token 无效: ${res.status}`);
  return res.json();
}

export async function getMyBots(ownerId: string): Promise<Bot[]> {
  const res = await fetch(`${API_BASE}/my?ownerId=${ownerId}`);
  if (!res.ok) throw new Error(`获取我的机器人失败: ${res.status}`);
  return res.json();
}

export async function getPublicBots(): Promise<Bot[]> {
  const res = await fetch(`${API_BASE}/public`);
  if (!res.ok) throw new Error(`获取公开机器人失败: ${res.status}`);
  return res.json();
}

export async function updateBot(botId: string, updates: Partial<Bot>): Promise<Bot> {
  const res = await fetch(`${API_BASE}/${botId}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(updates),
  });
  if (!res.ok) throw new Error(`更新机器人失败: ${res.status}`);
  return res.json();
}

export async function deleteBot(botId: string): Promise<void> {
  const res = await fetch(`${API_BASE}/${botId}`, { method: 'DELETE' });
  if (!res.ok) throw new Error(`删除机器人失败: ${res.status}`);
}

// ========== Chat ==========

export async function chatWithBot(
  botId: string,
  userId: string,
  conversationId: string,
  message: string
): Promise<ChatResponse> {
  const res = await fetch(`${API_BASE}/${botId}/chat`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ userId, conversationId, message }),
  });
  if (!res.ok) throw new Error(`AI 对话失败: ${res.status}`);
  return res.json();
}

export async function sendWebhookMessage(
  token: string,
  message: string,
  userId?: string,
  conversationId?: string
): Promise<{ reply: string }> {
  const res = await fetch(`${API_BASE}/webhook/${token}/receive`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ message, userId, conversationId }),
  });
  if (!res.ok) throw new Error(`Webhook 调用失败: ${res.status}`);
  return res.json();
}

// ========== Sessions ==========

export async function getSession(sessionId: string): Promise<BotSession> {
  const res = await fetch(`${API_BASE}/sessions/${sessionId}`);
  if (!res.ok) throw new Error(`获取会话失败: ${res.status}`);
  return res.json();
}

export async function getUserSessions(userId: string): Promise<BotSession[]> {
  const res = await fetch(`${API_BASE}/sessions/user/${userId}`);
  if (!res.ok) throw new Error(`获取用户会话失败: ${res.status}`);
  return res.json();
}

export async function endSession(sessionId: string, reason?: string): Promise<void> {
  const res = await fetch(`${API_BASE}/sessions/${sessionId}/end`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ reason: reason || 'USER_END' }),
  });
  if (!res.ok) throw new Error(`结束会话失败: ${res.status}`);
}

// ========== Stats ==========

export async function getBotStats(botId: string): Promise<BotStats> {
  const res = await fetch(`${API_BASE}/${botId}/stats`);
  if (!res.ok) throw new Error(`获取统计失败: ${res.status}`);
  return res.json();
}

export async function getBotLeaderboard(): Promise<Array<{
  botId: string;
  name: string;
  messageCount: number;
  sessionCount: number;
  totalTokensUsed: number;
}>> {
  const res = await fetch(`${API_BASE}/leaderboard`);
  if (!res.ok) throw new Error(`获取排行榜失败: ${res.status}`);
  return res.json();
}

// ========== Presets ==========

export async function getBotPresets(): Promise<BotPreset[]> {
  const res = await fetch(`${API_BASE}/presets`);
  if (!res.ok) throw new Error(`获取预设失败: ${res.status}`);
  return res.json();
}

// ========== Preset Config Templates ==========

export const PRESET_CONFIGS: Record<string, Partial<Bot>> = {
  OPENAI_GPT4: {
    botType: 'OPENAI',
    modelName: 'gpt-4',
    maxTokens: 8192,
    temperature: 0.7,
    systemPrompt: '你是一个有帮助的 AI 助手。',
  },
  OPENAI_GPT4_TURBO: {
    botType: 'OPENAI',
    modelName: 'gpt-4-turbo',
    maxTokens: 128000,
    temperature: 0.7,
    systemPrompt: '你是一个快速而准确的 AI 助手。',
  },
  OPENAI_GPT35: {
    botType: 'OPENAI',
    modelName: 'gpt-3.5-turbo',
    maxTokens: 16384,
    temperature: 0.7,
    systemPrompt: '你是一个经济实惠的 AI 助手。',
  },
  CLAUDE_OPUS: {
    botType: 'CLAUDE',
    modelName: 'claude-3-opus-20240229',
    maxTokens: 4096,
    temperature: 0.7,
    systemPrompt: 'You are a helpful AI assistant.',
  },
  CLAUDE_SONNET: {
    botType: 'CLAUDE',
    modelName: 'claude-3-sonnet-20240229',
    maxTokens: 4096,
    temperature: 0.7,
    systemPrompt: 'You are a fast and capable AI assistant.',
  },
  GEMINI_PRO: {
    botType: 'GEMINI',
    modelName: 'gemini-pro',
    maxTokens: 32768,
    temperature: 0.9,
    systemPrompt: '你是一个有用的 AI 助手。',
  },
  CUSTOM: {
    botType: 'CUSTOM',
    modelName: 'custom',
    maxTokens: 4096,
    temperature: 0.7,
    systemPrompt: 'Custom webhook bot.',
  },
};

export default {
  createBot,
  getBot,
  getBotByToken,
  getMyBots,
  getPublicBots,
  updateBot,
  deleteBot,
  chatWithBot,
  sendWebhookMessage,
  getSession,
  getUserSessions,
  endSession,
  getBotStats,
  getBotLeaderboard,
  getBotPresets,
  PRESET_CONFIGS,
};

File: im-mobile/lib/services/bot_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

/// AI 聊天机器人服务 - Flutter Dart API 封装
/// 支持 OpenAI/Claude/Gemini/Custom Webhook 多平台 AI 模型

class Bot {
  final String botId;
  final String name;
  final String? description;
  final String? avatarUrl;
  final String ownerId;
  final String botType; // OPENAI/CLAUDE/GEMINI/CUSTOM/LOCAL
  final String modelName;
  final String? apiKey;
  final String? apiBaseUrl;
  final String? webhookUrl;
  final String? webhookSecret;
  final String? systemPrompt;
  final int maxTokens;
  final double temperature;
  final String status; // ACTIVE/INACTIVE/DELETED
  final bool isPublic;
  final bool enableImageGen;
  final bool enableSpeechToText;
  final int rateLimit;
  final int sessionCount;
  final int messageCount;
  final int totalTokensUsed;
  final String? accessToken;
  final DateTime createdAt;
  final DateTime? updatedAt;
  final DateTime? lastActiveAt;

  Bot({
    required this.botId,
    required this.name,
    this.description,
    this.avatarUrl,
    required this.ownerId,
    required this.botType,
    required this.modelName,
    this.apiKey,
    this.apiBaseUrl,
    this.webhookUrl,
    this.webhookSecret,
    this.systemPrompt,
    required this.maxTokens,
    required this.temperature,
    required this.status,
    required this.isPublic,
    required this.enableImageGen,
    required this.enableSpeechToText,
    required this.rateLimit,
    required this.sessionCount,
    required this.messageCount,
    required this.totalTokensUsed,
    this.accessToken,
    required this.createdAt,
    this.updatedAt,
    this.lastActiveAt,
  });

  factory Bot.fromJson(Map<String, dynamic> json) {
    return Bot(
      botId: json['botId'],
      name: json['name'],
      description: json['description'],
      avatarUrl: json['avatarUrl'],
      ownerId: json['ownerId'],
      botType: json['botType'],
      modelName: json['modelName'],
      apiKey: json['apiKey'],
      apiBaseUrl: json['apiBaseUrl'],
      webhookUrl: json['webhookUrl'],
      webhookSecret: json['webhookSecret'],
      systemPrompt: json['systemPrompt'],
      maxTokens: json['maxTokens'] ?? 4096,
      temperature: (json['temperature'] ?? 0.7).toDouble(),
      status: json['status'] ?? 'INACTIVE',
      isPublic: json['isPublic'] ?? false,
      enableImageGen: json['enableImageGen'] ?? false,
      enableSpeechToText: json['enableSpeechToText'] ?? false,
      rateLimit: json['rateLimit'] ?? 60,
      sessionCount: json['sessionCount'] ?? 0,
      messageCount: json['messageCount'] ?? 0,
      totalTokensUsed: json['totalTokensUsed'] ?? 0,
      accessToken: json['accessToken'],
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
      lastActiveAt: json['lastActiveAt'] != null ? DateTime.parse(json['lastActiveAt']) : null,
    );
  }
}

class BotSession {
  final String sessionId;
  final String botId;
  final String userId;
  final String conversationId;
  final int contextTokens;
  final int turnCount;
  final int totalTokensUsed;
  final String status; // ACTIVE/ENDED/TIMEOUT
  final String? endReason;
  final String? promptVersion;
  final DateTime createdAt;
  final DateTime lastMessageAt;
  final DateTime? endedAt;

  BotSession({
    required this.sessionId,
    required this.botId,
    required this.userId,
    required this.conversationId,
    required this.contextTokens,
    required this.turnCount,
    required this.totalTokensUsed,
    required this.status,
    this.endReason,
    this.promptVersion,
    required this.createdAt,
    required this.lastMessageAt,
    this.endedAt,
  });

  factory BotSession.fromJson(Map<String, dynamic> json) {
    return BotSession(
      sessionId: json['sessionId'],
      botId: json['botId'],
      userId: json['userId'],
      conversationId: json['conversationId'],
      contextTokens: json['contextTokens'] ?? 0,
      turnCount: json['turnCount'] ?? 0,
      totalTokensUsed: json['totalTokensUsed'] ?? 0,
      status: json['status'] ?? 'ACTIVE',
      endReason: json['endReason'],
      promptVersion: json['promptVersion'],
      createdAt: DateTime.parse(json['createdAt']),
      lastMessageAt: DateTime.parse(json['lastMessageAt']),
      endedAt: json['endedAt'] != null ? DateTime.parse(json['endedAt']) : null,
    );
  }
}

class BotStats {
  final int totalSessions;
  final int activeSessions;
  final int totalTokens;
  final double avgTurns;
  final int totalMessages;
  final int totalTokensUsed;

  BotStats({
    required this.totalSessions,
    required this.activeSessions,
    required this.totalTokens,
    required this.avgTurns,
    required this.totalMessages,
    required this.totalTokensUsed,
  });

  factory BotStats.fromJson(Map<String, dynamic> json) {
    return BotStats(
      totalSessions: json['totalSessions'] ?? 0,
      activeSessions: json['activeSessions'] ?? 0,
      totalTokens: json['totalTokens'] ?? 0,
      avgTurns: (json['avgTurns'] ?? 0).toDouble(),
      totalMessages: json['totalMessages'] ?? 0,
      totalTokensUsed: json['totalTokensUsed'] ?? 0,
    );
  }
}

class BotService {
  static const String _baseUrl = '/api/bots';

  final http.Client _client;

  BotService({http.Client? client}) : _client = client ?? http.Client();

  // ========== Bot CRUD ==========

  Future<Bot> createBot({
    required String name,
    String? description,
    String? botType,
    String? modelName,
    required String ownerId,
  }) async {
    final response = await _client.post(
      Uri.parse(_baseUrl),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'name': name,
        'description': description,
        'botType': botType,
        'modelName': modelName,
        'ownerId': ownerId,
      }),
    );
    if (response.statusCode != 200) throw Exception('创建机器人失败: ${response.statusCode}');
    return Bot.fromJson(jsonDecode(response.body));
  }

  Future<Bot> getBot(String botId) async {
    final response = await _client.get(Uri.parse('$_baseUrl/$botId'));
    if (response.statusCode != 200) throw Exception('获取机器人失败: ${response.statusCode}');
    return Bot.fromJson(jsonDecode(response.body));
  }

  Future<List<Bot>> getMyBots(String ownerId) async {
    final response = await _client.get(Uri.parse('$_baseUrl/my?ownerId=$ownerId'));
    if (response.statusCode != 200) throw Exception('获取我的机器人失败: ${response.statusCode}');
    final List<dynamic> data = jsonDecode(response.body);
    return data.map((json) => Bot.fromJson(json)).toList();
  }

  Future<List<Bot>> getPublicBots() async {
    final response = await _client.get(Uri.parse('$_baseUrl/public'));
    if (response.statusCode != 200) throw Exception('获取公开机器人失败: ${response.statusCode}');
    final List<dynamic> data = jsonDecode(response.body);
    return data.map((json) => Bot.fromJson(json)).toList();
  }

  Future<Bot> updateBot(String botId, Map<String, dynamic> updates) async {
    final response = await _client.put(
      Uri.parse('$_baseUrl/$botId'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode(updates),
    );
    if (response.statusCode != 200) throw Exception('更新机器人失败: ${response.statusCode}');
    return Bot.fromJson(jsonDecode(response.body));
  }

  Future<void> deleteBot(String botId) async {
    final response = await _client.delete(Uri.parse('$_baseUrl/$botId'));
    if (response.statusCode != 204) throw Exception('删除机器人失败: ${response.statusCode}');
  }

  // ========== Chat ==========

  Future<Map<String, dynamic>> chatWithBot({
    required String botId,
    required String userId,
    required String conversationId,
    required String message,
  }) async {
    final response = await _client.post(
      Uri.parse('$_baseUrl/$botId/chat'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'userId': userId,
        'conversationId': conversationId,
        'message': message,
      }),
    );
    if (response.statusCode != 200) throw Exception('AI 对话失败: ${response.statusCode}');
    return jsonDecode(response.body);
  }

  // ========== Sessions ==========

  Future<BotSession> getSession(String sessionId) async {
    final response = await _client.get(Uri.parse('$_baseUrl/sessions/$sessionId'));
    if (response.statusCode != 200) throw Exception('获取会话失败: ${response.statusCode}');
    return BotSession.fromJson(jsonDecode(response.body));
  }

  Future<List<BotSession>> getUserSessions(String userId) async {
    final response = await _client.get(Uri.parse('$_baseUrl/sessions/user/$userId'));
    if (response.statusCode != 200) throw Exception('获取用户会话失败: ${response.statusCode}');
    final List<dynamic> data = jsonDecode(response.body);
    return data.map((json) => BotSession.fromJson(json)).toList();
  }

  Future<void> endSession(String sessionId, {String reason = 'USER_END'}) async {
    final response = await _client.post(
      Uri.parse('$_baseUrl/sessions/$sessionId/end'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'reason': reason}),
    );
    if (response.statusCode != 200) throw Exception('结束会话失败: ${response.statusCode}');
  }

  // ========== Stats ==========

  Future<BotStats> getBotStats(String botId) async {
    final response = await _client.get(Uri.parse('$_baseUrl/$botId/stats'));
    if (response.statusCode != 200) throw Exception('获取统计失败: ${response.statusCode}');
    return BotStats.fromJson(jsonDecode(response.body));
  }

  Future<List<Map<String, dynamic>>> getBotLeaderboard() async {
    final response = await _client.get(Uri.parse('$_baseUrl/leaderboard'));
    if (response.statusCode != 200) throw Exception('获取排行榜失败: ${response.statusCode}');
    final List<dynamic> data = jsonDecode(response.body);
    return data.cast<Map<String, dynamic>>();
  }

  // ========== Presets ==========

  Future<List<Map<String, String>>> getBotPresets() async {
    final response = await _client.get(Uri.parse('$_baseUrl/presets'));
    if (response.statusCode != 200) throw Exception('获取预设失败: ${response.statusCode}');
    final List<dynamic> data = jsonDecode(response.body);
    return data.cast<Map<String, String>>();
  }
}
